compiler/ballerina-compiler-api/src/main/java/org/ballerina/compiler/impl/NodeResolver.java

```diff
@@ -0,0 +1,1363 @@
+import org.wso2.ballerinalang.compiler.tree.statements.BLangRollback;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangTransaction;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangTupleDestructure;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangTupleVariableDef;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangWhile;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangWorkerSend;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangXMLNSStatement;
+import org.wso2.ballerinalang.compiler.tree.types.BLangArrayType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangBuiltInRefTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangConstrainedType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangErrorType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangFiniteTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangFunctionTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangIntersectionTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangLetVariable;
+import org.wso2.ballerinalang.compiler.tree.types.BLangObjectTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangRecordTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangStreamType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangTableTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangTupleTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangUnionTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangUserDefinedType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangValueType;
+import org.wso2.ballerinalang.compiler.util.diagnotic.DiagnosticPos;
+
+import java.util.List;
+
+/**
+ * Finds the enclosing AST node for the given position.
+ *
+ * @since 2.0.0
+ */
+class NodeResolver extends BLangNodeVisitor {
+
+    private LinePosition cursorPos;
+    private BLangNode enclosingNode;
+
+    BLangNode lookup(BLangCompilationUnit unit, LinePosition cursorPos) {
+        this.cursorPos = cursorPos;
+        this.enclosingNode = null;
+
+        for (TopLevelNode node : unit.topLevelNodes) {
+            if (!PositionUtil.withinBlock(this.cursorPos, node.getPosition())) {
+                continue;
+            }
+
+            ((BLangNode) node).accept(this);
+        }
+
+        return this.enclosingNode;
+    }
+
+    private void lookupNodes(List<? extends BLangNode> nodes) {
+        for (BLangNode node : nodes) {
+            if (!PositionUtil.withinBlock(this.cursorPos, node.pos)) {
+                continue;
+            }
+
+            node.accept(this);
+            // TODO: Check whether we can return from here.
+        }
+    }
+
+    private void lookupNode(BLangNode node) {
+        if (node == null) {
+            return;
+        }
+
+        if (PositionUtil.withinBlock(this.cursorPos, node.pos)) {
```

Comment:
is it possible to merge the two if checks to one and do an early return?

Reply:
Can merge. But it'd still be the same I guess